File: java-mvc/clean-architecture/src/main/java/com/ecommerce/clean/usecase/GetProductUseCase.java
package com.ecommerce.clean.usecase;

import com.ecommerce.clean.entity.Product;
import com.ecommerce.clean.usecase.port.ProductGateway;

import java.util.Optional;
import java.util.UUID;

/**
 * USE CASE: Get a single product by ID.
 */
public class GetProductUseCase {

    private final ProductGateway productGateway;

    public GetProductUseCase(ProductGateway productGateway) {
        this.productGateway = productGateway;
    }

    public Optional<Product> execute(UUID id) {
        return productGateway.findById(id);
    }
}

File: nodejs-mvc/clean-architecture/src/usecases/GetProductUseCase.ts
import { Product } from '../entities/Product';
import { ProductGateway } from './interfaces/ProductGateway';
import { ProductOutput } from './dto/ProductDTO';
import { ProductNotFoundError } from './errors/UseCaseErrors';

/**
 * @layer Use Cases
 * @description Retrieves a single product by ID.
 */
export class GetProductUseCase {
  constructor(private readonly gateway: ProductGateway) {}

  async execute(id: string): Promise<ProductOutput> {
    const product = await this.gateway.findById(id);
    if (!product) {
      throw new ProductNotFoundError(id);
    }
    return this.toOutput(product);
  }

  private toOutput(product: Product): ProductOutput {
    return {
      id: product.id,
      name: product.name,
      description: product.description,
      price: product.price,
      category: product.category,
      stockQuantity: product.stockQuantity,
      sku: product.sku,
      active: product.active,
      createdAt: product.createdAt.toISOString(),
      updatedAt: product.updatedAt.toISOString(),
    };
  }
}
